core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala

```diff
@@ -51,27 +54,81 @@ class LogDirFailureTest extends IntegrationTestHarness {
   @Before
   override def setUp() {
     super.setUp()
-    TestUtils.createTopic(zkUtils, topic, 1, 2, servers = servers)
+    TestUtils.createTopic(zkUtils, topic, 12, serverCount, servers = servers)
   }
 
   @Test
   def testIOExceptionDuringLogRoll() {
-    testProduceAfterLogDirFailure(Roll)
+    testProduceAfterLogDirFailureOnLeader(Roll)
   }
 
   @Test
   def testIOExceptionDuringCheckpoint() {
-    testProduceAfterLogDirFailure(Checkpoint)
+    testProduceAfterLogDirFailureOnLeader(Checkpoint)
+  }
+
+  @Test
+  def testReplicaFetcherThreadAfterLogDirFailureOnFollower() {
+    val consumer = consumers.head
+    subscribeAndWaitForAssignment(topic, consumer)
+    val producer = producers.head
+    val partition = new TopicPartition(topic, 0)
+
+    val partitionInfo = producer.partitionsFor(topic).asScala.find(_.partition() == 0).get
+    val leaderServerId = partitionInfo.leader().id()
+    val followerId = partitionInfo.replicas().map(_.id()).find(_ != leaderServerId).get
+    val followerServer = servers.find(_.config.brokerId == followerId).get
+
+
+    val running = new AtomicBoolean(true)
+    @volatile var numMessages = 0
+    val thread = new Thread() {
+      override def run(): Unit = {
+        val producer = TestUtils.createNewProducer(
+          TestUtils.getBrokerListStrFromServers(servers, protocol = securityProtocol),
+          securityProtocol = securityProtocol,
+          trustStoreFile = trustStoreFile,
+          retries = 5,
+          requestTimeoutMs = 2000,
+          acks = 1
+        )
+
+        while (running.get()) {
+          producer.send(new ProducerRecord(topic, s"xxxxxxxxxxxxxxxxxxxx-$numMessages".getBytes))
+          numMessages += 1
+        }
+        producer.close()
+      }
+    }
+
+    thread.start()
+    // Wait for some messages to be produced
+    Thread.sleep(1000)
+
+    // Make log directory of the partition on the follower broker offline
+    val replica = followerServer.replicaManager.getReplicaOrException(partition)
+    val logDir = replica.log.get.dir.getParentFile
+    followerServer.replicaManager.handleLogDirFailure(logDir.getAbsolutePath)
```

Comment:
I am wondering if we really need to run a producer/consumer in this test. Even w/o data, the follower will still be issuing fetch requests. handleLogDirFailure() actually removes the replica from the follower. To test this more accurately, perhaps we can add a method in ReplicaManger that just moves a partition to offline state and call it here?

Reply:
Yeah I think we can do that to catch the specific bug here. The suggested approach is simpler. On the other hand, the current approach seems more general because it simulates log directory failure instead of putting a specific record to offline state. Thus there seems to be pros/cons of each approach. I can implement the suggested approach if you recommend so.

Reply:
@junrao I have updated the patch as suggested. The thread is removed from the test and a method is added to ReplicaManger to move a partition to offline state.